File: qqbot-runtime/src/main/java/com/mieai/qqbot/runtime/supervisor/BotRuntimeStatus.java
package com.mieai.qqbot.runtime.supervisor;

import com.mieai.qqbot.domain.bot.BotId;
import com.mieai.qqbot.domain.bot.BotRevision;
import java.time.Instant;
import java.util.Objects;
import java.util.Optional;

/** Immutable diagnostic snapshot for one configured bot. */
public record BotRuntimeStatus(
        BotId botId,
        BotRevision configurationRevision,
        boolean desiredEnabled,
        BotRuntimeState state,
        Instant stateChangedAt,
        Optional<Instant> startedAt,
        Optional<Instant> readyAt,
        Optional<Instant> lastHeartbeatAt,
        Optional<Instant> lastDispatchAt,
        Optional<BotSessionSnapshot> session,
        long reconnectCount,
        Optional<BotRuntimeFailure> lastFailure,
        Optional<Instant> lastFailureAt) {

    public BotRuntimeStatus {
        Objects.requireNonNull(botId, "botId must not be null");
        Objects.requireNonNull(configurationRevision, "configurationRevision must not be null");
        Objects.requireNonNull(state, "state must not be null");
        Objects.requireNonNull(stateChangedAt, "stateChangedAt must not be null");
        startedAt = copy(startedAt, "startedAt");
        readyAt = copy(readyAt, "readyAt");
        lastHeartbeatAt = copy(lastHeartbeatAt, "lastHeartbeatAt");
        lastDispatchAt = copy(lastDispatchAt, "lastDispatchAt");
        session = copy(session, "session");
        if (reconnectCount < 0L) {
            throw new IllegalArgumentException("reconnectCount must not be negative");
        }
        lastFailure = copy(lastFailure, "lastFailure");
        lastFailureAt = copy(lastFailureAt, "lastFailureAt");
        if (lastFailure.isPresent() != lastFailureAt.isPresent()) {
            throw new IllegalArgumentException(
                    "lastFailure and lastFailureAt must either both be present or both be empty");
        }
    }

    private static <T> Optional<T> copy(Optional<T> value, String name) {
        Objects.requireNonNull(value, name + " must not be null");
        return value.map(Objects::requireNonNull);
    }
}

File: qqbot-runtime/src/main/kotlin/com/mieai/qqbot/runtime/supervisor/BotRuntimeStatus.kt
package com.mieai.qqbot.runtime.supervisor

import com.mieai.qqbot.domain.bot.BotId
import com.mieai.qqbot.domain.bot.BotRevision
import java.time.Instant

data class BotRuntimeStatus(
    val botId: BotId,
    val configurationRevision: BotRevision,
    val desiredEnabled: Boolean,
    val state: BotRuntimeState,
    val stateChangedAt: Instant,
    val startedAt: Instant?,
    val readyAt: Instant?,
    val lastHeartbeatAt: Instant?,
    val lastDispatchAt: Instant?,
    val session: BotSessionSnapshot?,
    val reconnectCount: Long,
    val lastFailure: BotRuntimeFailure?,
    val lastFailureAt: Instant?,
) {
    init {
        require(reconnectCount >= 0) { "reconnectCount must not be negative" }
        require((lastFailure == null) == (lastFailureAt == null)) {
            "lastFailure and lastFailureAt must be present together"
        }
    }
}
